geomesa-tools/src/main/scala/org/locationtech/geomesa/tools/ingest/AbstractIngest.scala

```diff
@@ -0,0 +1,235 @@
+
+  /**
+   * Run a distributed ingestion
+   *
+   * @param statusCallback for reporting status
+   * @return (success, failures) counts
+   */
+  def runDistributedJob(statusCallback: (Float, Long, Long, Boolean) => Unit): (Long, Long)
+
+  val ds = DataStoreFinder.getDataStore(dsParams)
+
+  // (progress, start time, pass, fail, done)
+  val statusCallback: (Float, Long, Long, Long, Boolean) => Unit =
+    if (dsParams.get(AccumuloDataStoreParams.mockParam.getName).exists(_.toBoolean)) {
+      val progress = printProgress(System.err, buildString('\u26AC', 60), ' ', _: Char) _
+      var state = false
+      (p, s, pass, fail, d) => {
+        state = !state
+        if (state) progress('\u15e7')(p, s, pass, fail, d) else progress('\u2b58')(p, s, pass, fail, d)
+      }
+    } else {
+      printProgress(System.err, buildString(' ', 60), '\u003d', '\u003e')
+    }
+
+  /**
+   * Main method to kick off ingestion
+   */
+  override def run(): Unit = {
+    beforeRunTasks()
+    val distPrefixes = Seq("hdfs://", "s3n://", "s3a://")
+    if (distPrefixes.exists(inputs.head.toLowerCase.startsWith)) {
+      logger.info("Running ingestion in distributed mode")
+      runDistributed()
+    } else {
+      logger.info("Running ingestion in local mode")
+      runLocal()
+    }
+    ds.dispose()
+  }
+
+  private def runLocal(): Unit = {
+
+    // Global failure shared between threads
+    val (written, failed) = (new AtomicLong(0), new AtomicLong(0))
+
+    val bytesRead = new AtomicLong(0L)
+
+    class LocalIngestWorker(file: File) extends Runnable {
+      override def run(): Unit = {
+        try {
+          // make feature writer lazy so that we can create the schema based off the input file
+          lazy val fw = ds.getFeatureWriterAppend(typeName, Transaction.AUTO_COMMIT)
+          val converter = createLocalConverter(file, failed)
+          // count the raw bytes read from the file, as that's what we based our total on
+          val countingStream = new CountingInputStream(new FileInputStream(file))
+          val is = PathUtils.handleCompression(countingStream, file.getPath)
+          try {
+            converter.convert(is).foreach { sf =>
+              val toWrite = fw.next()
+              toWrite.setAttributes(sf.getAttributes)
+              toWrite.getIdentifier.asInstanceOf[FeatureIdImpl].setID(sf.getID)
+              toWrite.getUserData.putAll(sf.getUserData)
+              toWrite.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+              try {
+                fw.write()
+                written.incrementAndGet()
+              } catch {
+                case e: Exception =>
+                  logger.error(s"Failed to write '${DataUtilities.encodeFeature(toWrite)}'", e)
+                  failed.incrementAndGet()
+              }
+              bytesRead.addAndGet(countingStream.getCount)
+              countingStream.resetCount()
+            }
+          } finally {
```

Comment:
It's a little gross but I got an exception here in this try from the AvroDataFile reader and it was swallowed bc of IOUtils.closeQuietly(fw)...the lazy eval hits in the finally block causing an exception which gets caught and the original is lost!